crates/viewer/re_space_view_graph/src/view.rs

```diff
@@ -147,122 +148,108 @@ Display a graph of nodes and edges.
         let edge_data = &system_output.view_systems.get::<EdgesVisualizer>()?.data;
 
         let graphs = merge(node_data, edge_data)
-            .map(|(ent, nodes, edges)| (ent, Graph::new(nodes, edges)))
+            .map(|(ent, nodes, edges)| Graph::new(ui, ent.clone(), nodes, edges))
             .collect::<Vec<_>>();
 
-        // We could move this computation to the visualizers to improve
-        // performance if needed.
-        let discriminator = {
-            let mut hasher = ahash::AHasher::default();
-            graphs.hash(&mut hasher);
-            Discriminator::new(hasher.finish())
-        };
-
         let state = state.downcast_mut::<GraphSpaceViewState>()?;
 
         let bounds_property = ViewProperty::from_archetype::<VisualBounds2D>(
             ctx.blueprint_db(),
             ctx.blueprint_query,
             query.space_view_id,
         );
-
         let bounds: blueprint::components::VisualBounds2D =
             bounds_property.component_or_fallback(ctx, self, state)?;
 
-        let layout_was_empty = state.layout_state.is_none();
-        let layout = state
-            .layout_state
-            .get(discriminator, graphs.iter().map(|(_, graph)| graph));
-
-        let mut needs_remeasure = false;
-
-        state.world_bounds = Some(bounds);
-        let bounds_rect: egui::Rect = bounds.into();
-
-        let mut scene_builder = SceneBuilder::from_world_bounds(bounds_rect);
-
-        if state.show_debug {
-            scene_builder.show_debug();
-        }
-
-        let (new_world_bounds, response) = scene_builder.add(ui, |mut scene| {
-            for (entity, graph) in &graphs {
-                // We use the following to keep track of the bounding box over nodes in an entity.
-                let mut entity_rect = egui::Rect::NOTHING;
-
-                let ent_highlights = query.highlights.entity_highlight((*entity).hash());
+        let view_rect = ui.max_rect();
+
+        let request = LayoutRequest::from_graphs(graphs.iter());
+        let layout = state.layout_state.get(request);
+
+        let (resp, new_bounds) = zoom_pan_area(
+            ui,
+            view_rect,
+            bounds.into(),
+            egui::Id::new(query.space_view_id),
+            |ui| {
+                let mut world_bounding_rect = egui::Rect::NOTHING;
+
+                for graph in &graphs {
+                    let entity_path = graph.entity();
+                    let entity_highlights = query.highlights.entity_highlight(entity_path.hash());
+
+                    // For now we compute the entity rectangles on the fly.
+                    let mut current_rect = egui::Rect::NOTHING;
+
+                    for node in graph.nodes() {
```

Comment:
This function is becoming both long and deep. Any chance we can break out one of the inner scopes to its own function?

Reply:
I will try to break it down a bit.